test/distributed/org/apache/cassandra/distributed/test/log/PlacementSimulator.java

```diff
@@ -985,20 +977,84 @@ public Diff<T> invert()
         }
     }
 
-    public static String diffsToString(Map<Range, Diff<Node>> placements)
+    public static Diff<Replica> additionsAndTransientToFull(Diff<Replica> unfiltered)
+    {
+        if (unfiltered.additions.isEmpty())
+            return null;
+        List<Replica> additions = new ArrayList<>(unfiltered.additions.size());
+        List<Replica> removals = new ArrayList<>(unfiltered.additions.size());
+        for (Replica added : unfiltered.additions)
+        {
+            // Include any new FULL replicas here. If there exists the removal of a corresponding Transient
+            // replica (i.e. a switch from T -> F), add that too. We want T -> F transitions to happen early
+            // in a multi-step operation, at the same time as new write replicas are added.
+            if (added.isFull())
+            {
+                additions.add(added);
+                for (Replica removed : unfiltered.removals)
+                    if (removed.node().equals(added.node()) && removed.isTransient())
+                        removals.add(removed);
+            }
+            else
+            {
+                // Conversely, when a replica transitions from F -> T, it's enacted late in a multi-step operation.
+                // So only include TRANSIENT additions if there is no removal of a corresponding FULL replica.
+                boolean include = unfiltered.removals.stream()
+                                                     .noneMatch(removed -> removed.node().equals(added.node())
+                                                                           && removed.isFull());
+                if (include)
+                    additions.add(added);
+            }
+        }
+        return new Diff<>(additions, removals);
+    }
+
+    public static Diff<Replica> removalsAndFullToTransient(Diff<Replica> unfiltered)
+    {
+        if (unfiltered.removals.isEmpty())
+            return null;
+        List<Replica> additions = new ArrayList<>(unfiltered.removals.size());
+        List<Replica> removals = new ArrayList<>(unfiltered.removals.size());
+        for (Replica removed : unfiltered.removals)
+        {
+            // Include any new FULL replicas here. If there exists the removal of a corresponding Transient
+            // replica (i.e. a switch from T -> F), add that too. We want T -> F transitions to happen early
+            // in a multi-step operation, at the same time as new write replicas are added.
+            if (removed.isFull())
+            {
+                removals.add(removed);
+                for (Replica added : unfiltered.additions)
```

Comment:
Nit: I think same comment applies then to the above `additionsAndTransientToFull` code.